src/sentry/api/endpoints/artifact_lookup.py

```diff
@@ -0,0 +1,336 @@
+            limit=10,
+        )
+        if rate_limited:
+            logger.info(
+                "notification.rate_limited",
+                extra={"project_id": project.id, "file_id": file_id},
+            )
+            return HttpResponse({"Too many download requests"}, status=429)
+
+        file = File.objects.filter(id=file_id).first()
+
+        if file is None:
+            raise Http404
+
+        try:
+            fp = file.getfile()
+            response = StreamingHttpResponse(
+                iter(lambda: fp.read(4096), b""), content_type="application/octet-stream"
+            )
+            response["Content-Length"] = file.size
+            return response
+        except OSError:
+            raise Http404
+
+    def get(self, request: Request, project: Project) -> Response:
+        """
+        List a Project's Individual Artifacts or Bundles
+        ````````````````````````````````````````
+
+        Retrieve a list of individual artifacts or artifact bundles for a given project.
+
+        :pparam string organization_slug: the slug of the organization to query.
+        :pparam string project_slug: the slug of the project to query.
+        :qparam string debug_id: If set, will query and return all the artifact
+                                  bundles that match one of the given `debug_id`s.
+        :qparam string url: If set, will query and return all the individual
+                             artifacts, or artifact bundles that contain files
+                             that match the `url`. This is using a substring-match.
+        :qparam string release: Used in conjunction with `url`.
+        :qparam string dist: Used in conjunction with `url`.
+
+        :auth: required
+        """
+
+        if request.GET.get("download") is not None:
+            if has_download_permission(request, project):
+                return self.download_file(request.GET.get("download"), project)
+            else:
+                return Response(status=403)
+
+        debug_ids = []
+        for debug_id in request.GET.getlist("debug_id"):
+            try:
+                debug_ids.append(normalize_debug_id(debug_id))
+            except SymbolicError:
+                pass
+
+        urls = request.GET.getlist("url")
+        release_name = request.GET.get("release")
+        dist_name = request.GET.get("dist")
+
+        # We want to have:
+        # - The (minimal?) `Set` of artifact bundles that include the files we are
+        # looking for
+        # - Any individual file that might be left over afterwards
+
+        # TODO: Possibly use the algorithm sketched up here:
+        # https://github.com/getsentry/sentry/pull/45697#issuecomment-1466389132
+        # That would narrow down our set of bundles to the minimum set that covers
+        # the file names we are querying for, and also leave us with the remaining
+        # set of file names that are not covered by any bundle, to look up below
```

Comment:
I'm having a hard time understanding where the referenced algorithm would hook in here. Would it be used in `collect_artifact_bundles_containing_debug_ids`?

Reply:
The algorithm will be used for resolving `collect_release_artifact_bundles_containing_urls`.

Reply:
With the goal of adding the fewest amount of bundle ids possible to the set.

Reply:
I believe so, yes. Though I haven’t spent too much time considering that algorithm.